LoopKit/HealthKitSampleStore.swift

```diff
@@ -141,42 +143,91 @@ public class HealthKitSampleStore {
+
+    func queryAnchorDidChange() {
+        // Subclasses can override
+    }
 
     /// Called in response to an update by the observer query
     ///
     /// - Parameters:
     ///   - query: The query which triggered the update
     ///   - error: An error during the update, if one occurred
-    internal func observeUpdates(to query: HKObserverQuery, error: Error?) {
-        guard error == nil else {
-            log.error("%@ notified with changes with error: %{public}@", query, String(describing: error))
+    internal final func observeUpdates(to query: HKObserverQuery, completionHandler: HKObserverQueryCompletionHandler?, error: Error?) {
+        if let error = error {
+            self.log.error("%@ notified with changes with error: %{public}@", query, String(describing: error))
+            completionHandler?()
             return
         }
 
-        let anchoredObjectQuery = HKAnchoredObjectQuery(
-            type: self.sampleType,
-            predicate: query.predicate,
-            anchor: self.queryAnchor,
-            limit: HKObjectQueryNoLimit
-        ) { (query, newSamples, deletedSamples, anchor, error) in
-            self.log.debug("%@: new: %d deleted: %d anchor: %@ error: %@", #function, newSamples?.count ?? 0, deletedSamples?.count ?? 0, String(describing: anchor), String(describing: error))
+        // Hold the completion handler (calling any existing ones) until our next anchored object query returns
+        self.observerQueryCompletionHandler = completionHandler
 
-            if let error = error {
-                self.log.error("%@: error executing anchoredObjectQuery: %@", String(describing: type(of: self)), error.localizedDescription)
-            }
+        let queryAnchor = self.queryAnchor
+        log.default("%@ notified with changes. Fetching from: %{public}@", query, queryAnchor.map(String.init(describing:)) ?? "0")
+
+        createAnchoredObjectQuery(predicate: query.predicate, anchor: queryAnchor)
+    }
 
-            self.processResults(from: query, added: newSamples ?? [], deleted: deletedSamples ?? [], error: error)
-            self.queryAnchor = anchor
+    private final func anchoredObjectQueryResultsHandler(query: HKAnchoredObjectQuery, newSamples: [HKSample]?, deletedSamples: [HKDeletedObject]?, anchor: HKQueryAnchor?, error: Error?) {
+        log.default("anchoredObjectQuery.resultsHandler: new: %{public}d deleted: %{public}d anchor: %{public}@ error: %{public}@", newSamples?.count ?? 0, deletedSamples?.count ?? 0, String(describing: anchor), String(describing: error))
+
+        guard let anchor = anchor else {
+            // Clear any existing completion handler (after calling it)
+            self.observerQueryCompletionHandler = nil
+            return
         }
 
-        healthStore.execute(anchoredObjectQuery)
+        processResults(from: query, added: newSamples ?? [], deleted: deletedSamples ?? [], anchor: anchor) { didSucceed in
```

Comment:
Looks like `error` is ignored?  Is that intentional?

Reply:
We log above if there is an error, and we will not proceed if anchor is nil.